locator/tablets.cc

```diff
@@ -959,10 +1014,69 @@ std::unordered_set<sstring> tablet_aware_replication_strategy::recognized_tablet
     return opts;
 }
 
+namespace {
+
+// Create a simple map [data center -> index]. This index is used as position index in the small vector
+// in the tablet_effective_replication_map::_tablet_to_dc_replication_factor_map.
+[[nodiscard]] tablet_effective_replication_map::datacenter_to_index_map build_datacenter_map(const auto& data_centers_list) {
+    uint32_t datacenter_index = 0;
+    tablet_effective_replication_map::datacenter_to_index_map datacenter_map;
+    datacenter_map.reserve(data_centers_list.size());
+    for (const sstring& datacenter : data_centers_list) {
+        datacenter_map.emplace(datacenter, datacenter_index);
+        ++datacenter_index;
+    }
+    return datacenter_map;
+}
+
+// Populate a map [tablet id -> replication factor list] with replication factors for each data center.
+// For each tablet index in the tablet map, it retrieves the current replica set using `get_endpoints_for_reading`
+// and counts replicas per data center.  Each data center's replication factor (RF) is saved under
+// its index according to the `datacenter_map`.
+future<tablet_effective_replication_map::tablet_to_datacenter_replication_factor_list_map> build_dc_replication_factor_map(const tablet_map& tablets,
+        const topology& topo, const tablet_effective_replication_map::datacenter_to_index_map& datacenter_map) {
+    tablet_effective_replication_map::tablet_to_datacenter_replication_factor_list_map tablet_to_dc_replication_factor_map;
+    tablet_to_dc_replication_factor_map.reserve(tablets.tablet_count());
+    for (size_t tablet_index = 0; tablet_index < tablets.tablet_count(); ++tablet_index) {
+        const tablet_replica_set& replicas = tablets.get_replicas_for_reading(tablet_id{tablet_index});
+        tablet_to_dc_replication_factor_map.emplace_back(datacenter_map.size(), 0);
+        auto& token_rf_list = tablet_to_dc_replication_factor_map.back();
+        for (const auto& node : replicas) {
+            const locator::node* node_ptr = topo.find_node(node.host);
+            if (node_ptr == nullptr) {
+                continue;
```

Comment:
This solves the problem of failing during early boot when we don't have topology, but it means we will compute wrong RF. This isn't safe.

Reply:
Hmm, the code in `tablet_effective_replication_map::to_replica_set` filers out the nonexistent replicas from the replica set that's returned from the `get_replicas_for_reading`, so my impression was that it's a normal situation.

Reply:
With my patch series that moves storage proxy to identify nodes by ids those hosts will not be filtered.

Reply:
> Hmm, the code in `tablet_effective_replication_map::to_replica_set` filers out the nonexistent replicas from the replica set that's returned from the `get_replicas_for_reading`, so my impression was that it's a normal situation.

For `get_replicas_for_reading()` it's safe because missing replicas will just not be contacted, same as if they are down. Dropping them for RF calculation is not safe, since we may incorrectly satisfy request with fewer replicas than is required for a quorum.

Reply:
> > Options I see to fix:
> > 
> > 1. Move Tablets initialization past the `group0` init
> 
> Won't work - we need them for erm building when tables are created.

May be we can move tablet creation later, when topology is already know.

> 
> > 2. Try moving topology initialization to the earlier point.
> 
> Yes, we can load group0 snapshot earlier. That seems the cleanest.

What do you mean by "load group0 snapshot". This is what group0 initialization does. So how is it different from 1?

Also we need to consider 3 cases here:
1 Normal boot - where group0 state is available early.
2. Upgrade - where it may start to be available while everything is initialized already (but this should not be a problem for tablets since there could not be tablets during upgrade) 
3. Bootstrap. As far as I see we call init_non_system_keyspaces before even joining the cluster, so it may be completely incorrect until snapshot is transferred and loading non existing snapshot earlier will not help.

Reply:
> > > Options I see to fix:
> > > 
> > > 1. Move Tablets initialization past the `group0` init
> > 
> > 
> > Won't work - we need them for erm building when tables are created.
> 
> May be we can move tablet creation later, when topology is already know.

That would be option 3, moving user table init past group0 init, unless we create tables without tablets, but then they're useless, and potentially unsafe, so it's still the same as option 3.

> 
> > > 2. Try moving topology initialization to the earlier point.
> > 
> > 
> > Yes, we can load group0 snapshot earlier. That seems the cleanest.
> 
> What do you mean by "load group0 snapshot". This is what group0 initialization does. So how is it different from 1?

Currently group0 init is done after messaging service init. We don't want to load user tables after messaging init, that's too late, because rpcs can already try to access tables. We could init group0 in memory state from local snapshot earlier, before user tables and before networking. This is option 2. 

> 
> Also we need to consider 3 cases here: 1 Normal boot - where group0 state is available early. 2. Upgrade - where it may start to be available while everything is initialized already (but this should not be a problem for tablets since there could not be tablets during upgrade) 3. Bootstrap. As far as I see we call init_non_system_keyspaces before even joining the cluster, so it may be completely incorrect until snapshot is transferred and loading non existing snapshot earlier will not help.

Bootstrap should be safe, since schema should be consistent with local group0 snapshot. No group0 means no user tables.

Reply:
> > > > Options I see to fix:
> > > > 
> > > > 1. Move Tablets initialization past the `group0` init
> > > 
> > > 
> > > Won't work - we need them for erm building when tables are created.
> > 
> > 
> > May be we can move tablet creation later, when topology is already know.
> 
> That would be option 3, moving user table init past group0 init, unless we create tables without tablets, but then they're useless, and potentially unsafe, so it's still the same as option 3.

Currently group0 init loads everything including tablets, no?

> 
> > > > 2. Try moving topology initialization to the earlier point.
> > > 
> > > 
> > > Yes, we can load group0 snapshot earlier. That seems the cleanest.
> > 
> > 
> > What do you mean by "load group0 snapshot". This is what group0 initialization does. So how is it different from 1?
> 
> Currently group0 init is done after messaging service init. We don't want to load user tables after messaging init, that's too late, because rpcs can already try to access tables. We could init group0 in memory state from local snapshot earlier, before user tables and before networking. This is option 2.

Sometimes it is impossible to create group0 before rpc: upgrade, bootstrap. And the code needs to handle all cases.

May be what we need is to move `group0_service.setup_group0_if_exist()` before `init_non_system_keyspaces`

Reply:
> Currently group0 init loads everything including tablets, no?
Yes. 
I'm currently testing moved group0 init before the `init_non_system_keyspaces`. I also had to move `cdc::generation_service`  since `setup_group0_if_exist` depends on it.  One test failed:  `topology_experimental_raft/test_crash_coordinator_before_streaming`

Reply:
Since tablets are initialized during group0 snapshot loading, probably we could remove the code below 
``` C++ 
            try {
                ss.local().load_tablet_metadata({}).get();
            } catch (...) {
                if (!cfg->maintenance_mode()) {
                    throw;
                }
                startlog.error("Failed to load tablet metadata (ignoring due to maintenance mode): {}", std::current_exception());
            }
```

Reply:
It looks like we need another PR, that will take care of the Tablets-Topology inconsistency in the ERM  in different cases.  For instance, when a node is bootstrapped and  a new snapshot is loaded in `group0_state_machine::transfer_snapshot` Tablets are updated earlier in the `schema_applier::merge_tables_and_views", and then in `add_column_family` a new ERM is created, that contains new Tablets map and old topology.

Reply:
> It looks like we need another PR, that will take care of the Tablets-Topology inconsistency in the ERM in different cases. For instance, when a node is bootstrapped and a new snapshot is loaded in `group0_state_machine::transfer_snapshot` Tablets are updated earlier in the `schema_applier::merge_tables_and_views", and then in `add_column_family` a new ERM is created, that contains new Tablets map and old topology.

We should fix that first, otherwise this PR will introduce a regression where we would fail to construct ERM in this case.

Reply:
> We should fix that first, otherwise this PR will introduce a regression where we would fail to construct ERM in this case.


I see it as 2 different issues:

1.  Incorrect Replication Factor calculation when tablets are in the transition state, that is addressed by this PR
2.  Occasionally invalid Effective Replication Map, containing Tablets map inconsistent with Topology. This PR does not introduce this issue, it just reveals an already existing issue. Tablets metadata should be updated only together with topology, and before creating the new ERM. 

So as these are 2 different problems IMO these should be fixed separately. This PR would dependent on the other PR.  @kbr-scylla mentioned that there is a PR https://github.com/scylladb/scylladb/pull/20853, that be solving the second issue. I'm looking at it now.

Reply:
Created https://github.com/scylladb/scylladb/issues/21856

Reply:
> So as these are 2 different problems IMO these should be fixed separately. This PR would dependent on the other PR. @kbr-scylla mentioned that there is a PR https://github.com/scylladb/scylladb/pull/20853

Ok.